[MVC-Step 1] 러너덕 미션 제출합니다.

Deocksoo opened the pull request:
안녕하세요 킹뽀대님, 이번 미션 리뷰이 러너덕입니다.
이번 미션에서는 기존 코드들이 변경되지 않도록 하면서 다른 프레임워크를 사용하도록 리펙토링하는 연습을 해보았습니다. 리펙토링하는 과정에서 이름을 참 많이 고민하게 되는데요, 적절한 이름들을 잘 골랐는지 모르겠습니다. 
테스트도 아직 부족한 부분이 있지만 일단 피드백을 한번 받아보고싶어서 요청 보내드려요. 많은 조언 부탁드립니다 :)

kmdngyu committed:
feat: 기능목록 추가

kmdngyu committed:
test: ReflectionTest의 showClass() 구현

kmdngyu committed:
feat: Junit3Test에서 test로 시작하는 메소드만 실행

kmdngyu committed:
feat: Junit4Test에서 MyTest 어노테이션 메소드만 실행

kmdngyu committed:
feat: private field에 값 할당

kmdngyu committed:
feat: 인자를 가지는 생성자의 인스턴스 생성

kmdngyu committed:
feat: Component Scan

kmdngyu committed:
docs: 2단계 기능목록

kmdngyu committed:
feat: HandlerExecution 구현

kmdngyu committed:
refactor: HandlerExecution test수정

kmdngyu committed:
refactor: loginForm에 annotation 기반 핸들링 적용

kmdngyu committed:
refactor: RequestMapping에서 method설정을 하지 않았을 때 request에 있는 메소드로 작동

모든 method에 대해서 처리할 수 있도록 수정합니다.

kmdngyu committed:
refactor: annotationHandlerMapping

kmdngyu committed:
refactor: DispatcherServlet 리팩토링

kmdngyu committed:
refactor: annotation기반 컨트롤러 기능 추가

kmdngyu committed:
refactor: code convention 맞춤

kmdngyu committed:
fix: login fail 버그 해결

redirect view로 되어 있어 제대로 페이지가 보이지 않는 문제를 jsp view로 변경해
해결하였습니다.

kmdngyu committed:
chore: HandlerAdapter관련 todo추가

Deocksoo committed:
Merge pull request #1 from Deocksoo/step1

step1

Deocksoo committed:
refactor: AnnotationHandlerMapping클래스 코드 정리

일부 메서드 네이밍을 변경하고 전체적인 코드를 보다 가독성 있게 리펙토링하였습니다.

Deocksoo committed:
refactor: ControllerAdapter를 추가하여 DispatcherServlet 리펙토링

기존의 Controller는 ViewName만 리턴합니다. 하지만 annotation 기반으로 작동하는 새로운
ExecutionHandler는 ModelAndView를 리턴합니다. 서로 다른 두 가지 리턴타입에 대해 DispatcherServlet에서
다른 처리를 해주어야 하므로 코드가 복잡해집니다. 그래서 ControllerAdapter를 추가하고 HandlerExecution과
ControllerAdapter가 Execution 인터페이스를 구현하도록 하여 코드 중복을 제거하였습니다.

Deocksoo committed:
refactor: nextstep.mvc.tobe 패키지 구조 변경

kingbbode commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -16,42 +19,47 @@
+        } catch (NotFoundHandlerException e) {
+            Controller controller = handlerMapping.getHandler(requestUri);
+            return new ControllerAdapter(controller);
         }
```
1. try - catch 의 예외 모델을 사용한 분기처리를 하셨네요. NotFoundHandlerException 은 어느 레이어에게 전파하기 위해 설정된 예외인가요? 이 예외가 DispatcherServlet 에서 일관성을 가져갈 수 있는 형태인가요? 분기를 위한 처리라면 불필요하고 과하다고 생각이 듭니다.

2. 기존 코드는 전혀 손대지 않고 호환하도록 하려는 의도였을까요? Adapter 를 가져가는 계층이 모호한 것 같아요 :) DispatcherServlet이 레거시와 새로운 모델의 구현을 알도록 설계하는 것은 추후 확장에 큰 불편함으로 돌아올 것 같아요.

Deocksoo replied:
1. 이 부분에서 annotationHandlerMapping에서 handler를 찾지 못했을 경우에 null을 되돌려주는 것이 찜찜해서 분기를 위해 try-catch를 사용해 예외처리를 했어요. 하지만 말씀해주신 것처럼 분기를 위한 처리라 불필요한 예외인것 같아요. 어떻게 수정하는 것이 좋을지 조금 더 고민해보겠습니다!

2. 리펙토링할 때 제가 레거시 코드를 어디까지 수정해도 될지 감이 오지 않아서 Adapter를 애매하게 사용하게 된것 같아요. 이 부분을 수정해보고 싶은데 Adapter를 가져가는 계층이 모호하다는 말씀을 제가 잘 이해하지 못했어요. 이 부분을 혹시 조금만 더 부연해서 설명해주실 수 있을까요?

kingbbode commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -16,10 +25,73 @@ public AnnotationHandlerMapping(Object... basePackage) {
+            return;
+        }
+
+        connectHandler(annotation, execution);
```
requestMethods 가 1개이든 2개이든 존재하지 않든 일관된 작성이 가능할 것 같네요 :)
분기가 되어야할 것은 이 한 부분이지 않을까요?
```
requestMethods.length == 0 ? RequestMethod.values() : requestMethods
```

Deocksoo replied:
RequestMethod를 지정하지 않은(모든 RequestMethod를 지원하는) 컨트롤러 메소드를 등록할 때, 기존에 등록되어있는 컨트롤러 메서드가 있는지 확인하고 등록하는 과정이 없으면 앞서 ReqeustMethod를 지정해서 등록한 컨트롤러 메서드가 있는 경우 덮어 쓰게 되지 않을까요? 말씀해주신 것처럼 깔끔하게 처리하고 싶은데, 이 부분때문에 처리가 조금 까다로워요. (제가 제시해주신 방향을 잘 이해한것이 맞나요..?)

kingbbode commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,10 +1,39 @@
+    private Method method;
+
+    public HandlerExecution(Class clazz, Method method) {
+        this.instance = getInstance(clazz);
```
인스턴스를 HandlerExecution 마다 생성되지 않고 효율적으로 생성할 수 있지 않을까요?

Deocksoo replied:
HandlerExecution은 매핑시에 url과 RequestMethod에 따라 보통 하나씩만 생성하여 매핑하니 크게 효율이 떨어지지 않을 거라고 생각했는데, 좀더 생각해보니 하나의 컨트롤러에 메서드를 여러개 만들면 메서드마다 같은 컨트롤러가 중복해서 생성되면서 효율이 떨어지겠군요. 수정하겠습니다!

kingbbode commented on nextstep-mvc/src/test/java/nextstep/mvc/tobe/AnnotationHandlerMappingTest.java:
```diff
@@ -27,11 +36,53 @@ public void create_find() throws Exception {
+    @DisplayName("RequestMapping의 RequestMethod를 지정해주면 해당 Http method를 지원한다.")
+    public void post_method_in_requestMapping_annotation() throws Exception {
+        MockHttpServletRequest request = new MockHttpServletRequest("POST", "/one-method");
+        Method expectedMethod = getControllerMethod(MyController.class, "handleOneRequestMethod");
```
코드가 조금이라도 수정되면 테스트가 많이 깨지겠네요.
검증 방식과 기대 값에 대한 설정을 다시 고려해보는 것이 좋을 것 같습니다.

kingbbode reviewed (changes requested):
안녕하세요! 
몇 가지 코멘트 작성하였습니다 :) 
확인 부탁드립니다.

kingbbode commented:
테스트 대상과 검증 방식에 대한 고찰을 많이 해보셨으면 합니다.

모든 클래스를 테스트 하려다보니, 테스트가 너무 많은 구현을 알고 있게 되었다고 생각되네요.
무엇을 테스트해야하고 어떤 것을 검증해야할지 다시 한번 생각해보시는게 좋을 것 같습니다.

Deocksoo reviewed (commented):
좋은 리뷰 너무 감사합니다! 특히 테스트 코드를 작성하면서 찜찜했던 부분들이 많은데, 어떻게 테스트하는 것이 좋을지 더 고민해보겠습니다!

Deocksoo committed:
refactor: AnnotationHandlerMapping에서 Controller 객체 생성 및 주입 담당하도록 수정

Deocksoo committed:
refactor: ManualHandlerMapping과 Controller의 불필요한 코드 제거

리펙토링 과정에서 Annotation 기반 매핑으로 변경한 후에 남아있던 불필요한 코드들을
 제거합니다.

Deocksoo committed:
refactor: AnnotationHandlerMapping 리펙토링

ComponentScan 후에 매핑할 requestMethod 정하고, 정해진 requestMethod에 따라
일관적으로 Key와 execution을 매핑하도록 수정하였습니다.

Deocksoo committed:
refactor: DispatcherServlet에 HandlerMapping을 주입하는 방식으로 의존성 관리

DispatcherServlet이 구현체를 너무 많이 알고 있고, ControllerAdapter의 계층이 모호해지는 문제가
있어 HadlerMapping을 주입받도록 수정하였습니다.

Deocksoo committed:
refactor: 불필요한 테스트코드 제거, 테스트코드에서 리플렉션 사용 제거

Deocksoo committed:
refactor: ServletRequestHandler 인터페이스명을 HandlerAdapter로 변경

Deocksoo commented:
피드백 남겨주신 부분을 중심으로 코드를 정리해봤는데, 깔끔하게 정리하기가 쉽지 않네요. 아직 미흡한 부분들이 남아있는거 같아요, 많은 지적 부탁드립니다.
그리고 기존 테스트 코드에서 너무 많은 구현을 알고 있다는 말씀을 해주셔서 테스트 관련 코드들을 조금 찾아보면서 고민해보았는데요, 리플렉션을 사용해서 클래스 구현체에 직접 접근하는 코드들이나, HandlerExecutionTest와 같은 테스트들이 과했던 것이 아닌가 싶어 수정해보았습니다. 
제출이 조금 늦었지만 다시한번 리뷰 부탁드립니다, 감사합니다 :)

woowahankingbbode commented on nextstep-mvc/src/main/java/nextstep/mvc/DispatcherServlet.java:
```diff
@@ -1,57 +1,69 @@
-        } catch (Throwable e) {
-            logger.error("Exception : {}", e);
-            throw new ServletException(e.getMessage());
+            HandlerAdapter handler = getHandlerOf(request);
```
Handler 없이 HandlerAdapter 가 존재하는게 이상하네요 :)
Adapter 는 기존 결과물의 변경없이 새로운 구조와 연결이 될 수 있다는 의미를 가지고 있습니다.
의도하신 네이밍과 일치하는지 생각해보시면 좋을 것 같습니다!

woowahankingbbode commented on nextstep-mvc/src/main/java/nextstep/utils/LoggingUtils.java:
```diff
@@ -0,0 +1,16 @@
+import org.slf4j.Logger;
+
+public class LoggingUtils {
+    public static void logStackTrace(Logger logger, Exception e) {
```
이미 로거 구현체에서 하고 있는 일을 직접 구현하려고 하는 것 같아요.
로거 구현체의 사용법을 좀 더 알아보시는 것이 좋을 것 같아요.

Deocksoo replied:
이미 로거에 있는 기능이었네요..! 구현되어있는 기능을 사용하는 것이 더 좋을것 같아 수정했습니다 :)

woowahankingbbode commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/scanner/ControllerScanner.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+import java.util.Set;
+
+public class ControllerScanner {
```
좋은 분리네요 :)

woowahankingbbode reviewed (approved):
리뷰가 많이 늦었습니다!
1단계에서 필요한 내용은 모두 만족한 것 같아요.
2단계에서도 잘 부탁드려요! :)

woowahankingbbode commented:
처음 작성하신 형태의 추상화가 더 좋아보이네요 :)
계층이 모호하다고 말씀드린 어렵게 이야기된듯 하네요 ㅜㅜ 

새로운걸로 꽂아보고 안되면 기존 것으로 꼽아! 가 아니고 어뎁터를 가져갔다면 새로운 것과 기존 것을 구분하지 않아야한다는 의도로 말씀드렸는데 답변을 못했네요 ㅜㅜ

woowahankingbbode commented:
`ReqeustMethod를 지정해서 등록한 컨트롤러 메서드가 있는 경우 덮어 쓰게 되지 않을까요?`

위 부분은 RequestMethod 가 모두일 뿐만 아니라 모든 경우에 대해서도 확인이 필요한 부분이지 않을까요?

```
for (RequestMethod requestMethod : connectable) {
            handlerKey = new HandlerKey(uri, requestMethod);
            handlerExecutions.put(handlerKey, execution);
        }
```

제가 보기에는 이 단계에서 처리해야할 문제인 것 같아요 :) 이렇게 보면 `connectHandlerToAllConnectableRequestMethods` 과 `connectHandler` 이 중복된 일을 하고 있는게 느껴지실까요?

woowahankingbbode commented:
고민이 많으셨을텐데 코멘트 답글로 달렸을 때 이메일이 너무 많이와서 다 확인할 수가 없었네요  ㅜㅜ
다음에는 고민이 생기시면 슬랙 DM 으로 코멘트 달았어요! 라고 보내주시면 감사하겠습니다 ㅜㅜ
